client/securedrop_client/gui/widgets.py

```diff
@@ -424,6 +429,139 @@ def clear_message(self) -> None:
+        self.setObjectName("BatchActionWidget")
+
+        # Solid background colour
+        self.setAttribute(Qt.WA_StyledBackground, True)
+        layout = QHBoxLayout()
+        self.setLayout(layout)
+
+        self.toolbar = BatchActionToolbar()
+
+        layout.addWidget(self.toolbar)
+        layout.addStretch()
+
+    def setup(self, controller: Controller) -> None:
+        self.toolbar.setup(controller)
+
+
+class BatchActionToolbar(QToolBar):
+    """
+    A toolbar that contains batch actions (actions that target multiple
+    sources in the ConversationView, and therefore don't belong in the
+    individual conversation menu). Currently, this widget will hold the
+    "Delete Sources" (batch-delete) action.
+
+    For user-facing naming consistency, these items won't be called
+    "batch/bulk <delete>", but simply "<verb> <noun>s" (eg "Delete Sources"), where
+    the original nomenclature comes from the individual Source overflow QAction menu
+    items. Each item may have a tooltip, visible on hover, that provides a more
+    lengthy explanation (e.g., "Delete multiple source accounts").
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.setObjectName("BatchActionToolbar")
+        self.setContentsMargins(0, 0, 0, 0)
+
+        palette = QPalette()
+        palette.setBrush(
+            QPalette.Background, QBrush(Qt.NoBrush)
+        )  # This makes the widget transparent
+        self.setPalette(palette)
+
+        # Style and attributes
+        self.setMovable(False)
+        self.setToolButtonStyle(Qt.ToolButtonStyle.ToolButtonTextBesideIcon)
+        self.setSizePolicy(QSizePolicy.Preferred, QSizePolicy.Fixed)
+
+        self.delete_sources_action = QAction(
+            QIcon(load_image("delete_sources_toolbar_icon.svg")), _("DELETE SOURCES"), self
+        )
+        self.delete_sources_action.setObjectName("BatchActionButton")
+        self.delete_sources_action.setToolTip(
+            _(
+                "Delete selected source accounts. "
+                "Ctrl+click sources below to select multiple sources."
+            )
+        )
+        self.delete_sources_action.triggered.connect(self.on_action_triggered)
+        self.button = self.widgetForAction(self.delete_sources_action)
+
+        # Add spacer. "select all" checkbox can replace the spacer in future
+        spacer = QWidget()
+        spacer.setFixedSize(14, 14)  # sourcewidget spacer size
+        self.addWidget(spacer)
+        self.addAction(self.delete_sources_action)
+
+    def setup(self, controller: Controller) -> None:
+        self.controller = controller
+
+    def hide_action(self) -> None:
+        self.delete_sources_action.setVisible(False)
+
+    def show_action(self) -> None:
+        self.delete_sources_action.setVisible(True)
+
+    @pyqtSlot()
+    def on_action_triggered(self) -> None:
+        if self.controller.api is None:
+            self.controller.on_action_requiring_login()
```

Comment:
I agree and was irked by this.  I kept to this pattern because it's used elsewhere in the gui, but what I think should happen ultimately is a bigger refactor along these lines:
- child gui widgets don't have a reference to the controller
- when child gui widget clicks a button, it sends a request signal
- controller handles the signal and emits a response

That way it's none of the widget's business if the app is logged in or not, and it's easier to test and create test conditions.